pkg/solana/txm/txm.go

```diff
@@ -175,197 +179,215 @@ func (txm *Txm) run() {
-							txm.lggr.Warnw("error in adding retry transaction", "error", retryStoreErr, "id", msg.id)
-							return
-						}
-						if setErr := sigs.Set(count, retrySig); setErr != nil {
-							// this should never happen
-							txm.lggr.Errorw("INVARIANT VIOLATION", "error", setErr)
-						}
-						txm.lggr.Debugw("tx rebroadcast with bumped fee", "id", msg.id, "fee", getFee(count), "signatures", sigs.List())
-					}
-
-					// prevent locking on waitgroup when ctx is closed
-					wait := make(chan struct{})
-					go func() {
-						defer close(wait)
-						sigs.Wait(count) // wait until bump tx has set the tx signature to compare rebroadcast signatures
-					}()
-					select {
-					case <-ctx.Done():
-						return
-					case <-wait:
-					}
-
-					// this should never happen (should match the signature saved to sigs)
-					if fetchedSig, fetchErr := sigs.Get(count); fetchErr != nil || retrySig != fetchedSig {
-						txm.lggr.Errorw("original signature does not match retry signature", "expectedSignatures", sigs.List(), "receivedSignature", retrySig, "error", fetchErr)
-					}
-				}(shouldBump, bumpCount, currentTx)
-			}
+	// Set compute unit limit if specified
+	if msg.cfg.ComputeUnitLimit != 0 {
+		if err := fees.SetComputeUnitLimit(&newTx, fees.ComputeUnitLimit(msg.cfg.ComputeUnitLimit)); err != nil {
+			return solanaGo.Transaction{}, fmt.Errorf("failed to add compute unit limit instruction: %w", err)
+		}
+	}
 
-			// exponential increase in wait time, capped at 250ms
-			deltaT *= 2
-			if deltaT > MaxRetryTimeMs {
-				deltaT = MaxRetryTimeMs
+	// Set compute unit price (fee)
+	fee := fees.ComputeUnitPrice(
+		fees.CalculateFee(
+			msg.cfg.BaseComputeUnitPrice,
+			msg.cfg.ComputeUnitPriceMax,
+			msg.cfg.ComputeUnitPriceMin,
+			uint(retryCount), //nolint:gosec // reasonable number of bumps should never cause overflow
+		))
+	if err := fees.SetComputeUnitPrice(&newTx, fee); err != nil {
+		return solanaGo.Transaction{}, err
+	}
+
+	// Sign transaction
+	// NOTE: fee payer account is index 0 account. https://github.com/gagliardetto/solana-go/blob/main/transaction.go#L252
+	txMsg, err := newTx.Message.MarshalBinary()
+	if err != nil {
+		return solanaGo.Transaction{}, fmt.Errorf("error in MarshalBinary: %w", err)
+	}
+	sigBytes, err := txm.ks.Sign(ctx, msg.tx.Message.AccountKeys[0].String(), txMsg)
+	if err != nil {
+		return solanaGo.Transaction{}, fmt.Errorf("error in Sign: %w", err)
+	}
+	var finalSig [64]byte
+	copy(finalSig[:], sigBytes)
+	newTx.Signatures = append(newTx.Signatures, finalSig)
+
+	return newTx, nil
+}
+
+// retryTx contains the logic for retrying the transaction, including exponential backoff and fee bumping.
+// Retries until context cancelled by timeout or called externally.
+// It uses handleRetry helper function to handle each retry attempt.
+func (txm *Txm) retryTx(ctx context.Context, msg pendingTx, currentTx solanaGo.Transaction, sigs *signatureList) {
+	deltaT := 1 // initial delay in ms
```

Comment:
Isn't 1 ms too aggressive?
I assume just a round-trip time to Solana RPC call would be in order of 10s of ms atleast.
So after initial send, we may want to wait atleast 10 ms to give the initial tx some time to get confirmed.

Otherwise if say we have initial deltaT as 1ms, then next exponential backoffs are at 1, 2 ,4 8, 16, 31, 64, 128 ms.

But I see that the confirm logic, which checks Tx confirmation, has a polling that's 500 ms:
ConfirmPollPeriod:   config.MustNewDuration(500 * time.Millisecond)

Seems like we will retry many times even before trying to check confirmation once.

Reply:
This was logic that carried over from the existing code for Data Feeds. We chose to keep this the same so we don't introduce any regressions from DF. We could maybe make this configurable in the future especially if we need to reduce RPC calls but seems like this was out of necessity to make sure our transaction gets out in front of the current validator. The multinode integration probably helps a bit with that though so spamming this much maybe isn't needed anymore. I had the same thoughts as you initially but was reluctant to relax the existing code in case it introduced any unforeseen issues.

Reply:
We [decided here](https://chainlink-core.slack.com/archives/C04Q61M2MFT/p1733936968071019?thread_ts=1733426993.295679&cid=C04Q61M2MFT) to create a new ticket to make this retry configurable. We will keep the default values as they are for now. This will give us the flexibility to adjust them based on the use case. For example, we could relax these configurations for CCIP and make them more aggressive for DF